src/Parser/Parser.jl

```diff
@@ -0,0 +1,398 @@
+
+function read_four_hex_digits!(ps::ParserState)
+    local n::UInt16 = 0
+
+    for _ in 1:4
+        b = advance!(ps)
+        n = n << 4 + if isjsondigit(b)
+            b - DIGIT_ZERO
+        elseif LATIN_A ≤ b ≤ LATIN_F
+            b - (LATIN_A - @compat UInt8(10))
+        elseif LATIN_UPPER_A ≤ b ≤ LATIN_UPPER_F
+            b - (LATIN_UPPER_A - @compat UInt8(10))
+        else
+            _error(E_BAD_ESCAPE, ps)
+        end
+    end
+
+    n
+end
+
+function read_unicode_escape!(ps)
+    u1 = read_four_hex_digits!(ps)
+    if utf16_is_surrogate(u1)
+        skip!(ps, BACKSLASH)
+        skip!(ps, LATIN_U)
+        u2 = read_four_hex_digits!(ps)
+        utf16_get_supplementary(u1, u2)
+    else
+        @compat(Char(u1))
+    end
+end
+
+
+function parse_string(ps::ParserState)
+    b = UInt8[]
+    incr!(ps)  # skip opening quote
+    while true
+        c = advance!(ps)
+
+        if c == BACKSLASH
+            c = advance!(ps)
+            if c == LATIN_U  # Unicode escape
+                append!(b, string(read_unicode_escape!(ps)).data)
+            else
+                c = get(ESCAPES, c, 0x00)
+                c == 0x00 && _error(E_BAD_ESCAPE, ps)
+                push!(b, c)
+            end
+            continue
+        elseif c < SPACE
+            _error(E_BAD_CONTROL, ps)
+        elseif c == STRING_DELIM
+            return _String(b)
+        end
+
+        push!(b, c)
+    end
+end
+
+"""
+Return `true` if the given bytes vector, starting at `from` and ending at `to`,
+has a leading zero.
+"""
+function hasleadingzero(bytes::Vector{UInt8}, from::Int, to::Int)
+    c = bytes[from]
+    from + 1 < to && c == @compat(UInt8('-')) &&
+            bytes[from + 1] == DIGIT_ZERO && isjsondigit(bytes[from + 2]) ||
+    from < to && to > from + 1 && c == DIGIT_ZERO &&
+            isjsondigit(bytes[from + 1])
+end
+
+"""
+Parse a float from the given bytes vector, starting at `from` and ending at the
+byte before `to`. Bytes enclosed should all be ASCII characters.
+"""
+function float_from_bytes(bytes::Vector{UInt8}, from::Int, to::Int)
+    # The ccall is not ideal (Base.tryparse would be better), but it actually
+    # makes an 2× difference to performance
+    @static if VERSION ≥ v"0.4"
+        ccall(:jl_try_substrtod, Nullable{Float64},
+                (Ptr{UInt8}, Csize_t, Csize_t), bytes, from - 1, to - from + 1)
+    else
+        tryparse(Float64, Compat.ASCIIString(bytes[from:to]))
+    end
+end
+
+"""
+Parse an integer from the given bytes vector, starting at `from` and ending at
+the byte before `to`. Bytes enclosed should all be ASCII characters.
+"""
+function int_from_bytes(bytes::Vector{UInt8}, from::Int, to::Int)
+    @inbounds isnegative = bytes[from] == MINUS_SIGN ? (from += 1; true) : false
+    num = @compat Int64(0)
+    @inbounds for i in from:to
+        num = (@compat Int64(10)) * num +
+                @compat Int64(bytes[i] - DIGIT_ZERO)
```

Comment:
Julia uses `Int` as an alias for `Int32` on 32-bit systems and `Int64` on 64-bit systems.  Is there a reason we shouldn't follow that convention here?

(As noted before, the JSON spec doesn't actually define integers as part of the spec, only "numbers" (which are floats).  We could follow that, but some packages seem to depend on parsing ints.)